ingest_operations.py

```diff
@@ -336,90 +383,195 @@ async def approve_pending_users():
+    if not auth.is_site_admin(token):
+        return {"error": f"User not authorized to get preapproved users"}, 403
+
+    user_name = urllib.parse.unquote_plus(user_id)
+
+    response, status_code = authx.auth.get_preapproved_user_in_opa(user_name)
+    return response, status_code
+
+
+@app.route('/user/preapproved/<path:user_id>')
+def add_preapproved_user(user_id):
+    token = connexion.request.headers['Authorization'].split("Bearer ")[1]
+    if not auth.is_site_admin(token):
+        return {"error": f"User not authorized to add preapproved users"}, 403
+
+    user_name = urllib.parse.unquote_plus(user_id)
+
+    response, status_code = authx.auth.add_preapproved_user_in_opa(user_name)
+    return response, status_code
 
 
-@app.route('/user/<path:user_id>/authorize')
-def list_programs_for_user(user_id):
+@app.route('/user/preapproved/<path:user_id>')
+def remove_preapproved_user(user_id):
     token = connexion.request.headers['Authorization'].split("Bearer ")[1]
+    if not auth.is_site_admin(token):
+        return {"error": f"User not authorized to remove preapproved users"}, 403
+
+    user_name = urllib.parse.unquote_plus(user_id)
+
+    response, status_code = authx.auth.remove_preapproved_user_in_opa(user_name)
+    return response, status_code
+
+
+####
+# DAC authorization for users
+####
+
+@app.route('/user/<path:user_id>')
+def list_authz_for_user(user_id):
+    token = connexion.request.headers['Authorization'].split("Bearer ")[1]
+
     response = ""
     status_code = 0
+    if not authx.auth.is_action_allowed_for_program(token, method="GET", path=f"/ingest/user/{user_id}", program=None):
+        return {"error": "User not authorized to list programs for user"}, 403
+
     if user_id == "me":
-        # Grab the user's own authorization
-        response, status_code = list_programs_for_self(token)
-    else:
-        user_name = urllib.parse.unquote_plus(user_id)
-        response, status_code = auth.get_user_in_opa(user_name, token)
-        if status_code != 200:
-            return response, status_code
-        response = list(response["programs"].values())
-    print(response)
-    return {"results": response}, status_code
+        user_id = authx.auth.get_user_id(request)
+
+    user_result, status_code = authx.auth.get_user_in_opa(user_id)
+    if status_code != 200:
+        return user_result, status_code
+
+    user_result["site_roles"] = []
+    role_types, status_code = authx.auth.list_role_types_in_opa()
+    if status_code == 200:
+        for role_type in role_types:
+            users, status_code = authx.auth.get_role_type_in_opa(role_type)
+            if user_id in users[role_type]:
+                user_result["site_roles"].append(role_type)
+
+    user_result["program_authorizations"] = {}
+    opa_permissions, status_code = authx.auth.get_opa_permissions(bearer_token=token, user_token=user_result["userinfo"]["sample_jwt"])
```

Comment:
I'm a bit confused now; the return value of `ingest/user/me` is 200 if the user exists, 404 if the user doesn't exist, and then... how do I tell if I'm a pending user?

Reply:
You would try to request pending access again, `/ingest/pending/request`...if you get a 200, you're already pending.

Reply:
Can we get a non-destructive way of checking pending status? When the user enters the site we want to show them a different screen depending on whether or not they're already pending.

Reply:
I don't understand how it's destructive: if you repeatedly hit the pending/request endpoint, you just get repeated 200 OKs...nothing changes.

Reply:
![image](https://github.com/user-attachments/assets/fb2d58a6-a95c-42f0-a29f-3ba0df4221be)
When the user first visits the page, they should be given the screen on the left. When they click the button to POST to `/ingest/pending/request`, they should be given the screen on the right. From then onwards, whenever they check the page, they should immediately be given the screen on the right.

The problem I'd run into is that it's impossible to determine which screen the user should be given when they load up the page without inadvertently signing them up to be on the pending list, or via using some sort of cookie/local storage check (which would fail if they checked it in a different browser or on a different computer).

Reply:
When we last discussed this I had thought there was some GET endpoint with a different status code depending on the user's authorization state -- was it only the POST endpoint that was planned?

Reply:
Oh I see. I'll try to add another endpoint then.

Reply:
I added GET to /user/pending/{user_id}. "me" works to get the status of yourself. (had to add a bit in Opa; please pull the opa commit in https://github.com/CanDIG/CanDIGv2/pull/938 and recompose opa.